### PR TITLE
Use region heartbeat to determine whether recovery is finished ref #4242

### PR DESCRIPTION
**Pull request by v01dstar**
Signed-off-by: v01dstar <yang.zhang@pingcap.com>

### What problem does this PR solve?

This is part of https://github.com/tikv/tikv/issues/10483

### What is changed and how it works?

This PR makes PD check region info reported by region heartbeats to determine whether unsafe recovery plan has been executed, instead of the report sent back from TiKV, to reduce network usage.

### Check List

Tests

- Unit tests

Code changes

N/A

Side effects

N/A

Related changes

N/A

### Release note

```release-note
None
```

### CODE REVIEW

File: server/cluster/unsafe_recovery_controller.go [hunk 1]
```diff
@@ -473,7 +467,7 @@ func (u *unsafeRecoveryController) Show() []string {
 		status = append(status, "Execution progess:")
 		for storeID, applied := range u.executionResults {
 			if !applied {
-				status = append(status, strconv.FormatUint(storeID, 10)+"not yet applied, last report: "+getStoreDigest(u.executionReports[storeID]))
+				status = append(status, strconv.FormatUint(storeID, 10)+"not yet applied")
```
Comment (rleungx) [inline server/cluster/unsafe_recovery_controller.go:470]: How about using Sprintf?

File: server/cluster/unsafe_recovery_controller.go [hunk 2]
```diff
@@ -149,36 +142,37 @@ func (u *unsafeRecoveryController) reset() {
...
 	u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
 	u.executionResults = make(map[uint64]bool)
-	u.executionReports = make(map[uint64]*pdpb.StoreReport)
 	u.numStoresPlanExecuted = 0
 }
 
-func (u *unsafeRecoveryController) isPlanExecuted(storeID uint64, report *pdpb.StoreReport) bool {
-	targetRegions := make(map[uint64]*metapb.Region)
-	toBeRemovedRegions := make(map[uint64]bool)
+func (u *unsafeRecoveryController) isPlanExecuted(storeID uint64) bool {
```
Comment (rleungx) [inline server/cluster/unsafe_recovery_controller.go:148]: How do we know if there are some problems when executing the plan

**Review by Connor1996 — COMMENTED** [review]
Let's still keep store heartbeat to determine, but reduce extra peer report

**Comment by JmPotato** [comment]
Any updates on this?

**Pull request closed without merging**